lxd/container_put.go

```diff
@@ -39,6 +39,60 @@ func containerPut(d *Daemon, r *http.Request) Response {
 		return BadRequest(err)
 	}
 
+	oldLocalRootDevice := containerGetRootDiskDevice(c.LocalDevices())
+	newLocalRootDevice := containerGetRootDiskDevice(configRaw.Devices)
+
+	var profileRootDiskDevices []string
+	// Check whether the old profiles have a root disk device.
+	for _, pName := range c.Profiles() {
+		_, p, err := dbProfileGet(d.db, pName)
+		if err != nil {
+			continue
+		}
+
+		v := containerGetRootDiskDevice(p.Devices)
+		if !shared.StringInSlice(v["pool"], profileRootDiskDevices) {
+			profileRootDiskDevices = append(profileRootDiskDevices, v["pool"])
+		}
+	}
+
+	// Check whether the new profiles have a root disk device.
+	for _, pName := range configRaw.Profiles {
+		_, p, err := dbProfileGet(d.db, pName)
+		if err != nil {
+			continue
+		}
+
+		v := containerGetRootDiskDevice(p.Devices)
+		if !shared.StringInSlice(v["pool"], profileRootDiskDevices) {
+			profileRootDiskDevices = append(profileRootDiskDevices, v["pool"])
+		}
+	}
+
+	// We know there are two different root disk devices specified in the
+	// profiles the container is supposed to be attached to. We
+	// shouldn't allow that to happen as this would be nasty to deal with
+	// when the container removes it's root disk device.
+	if len(profileRootDiskDevices) > 1 && newLocalRootDevice == nil || (len(profileRootDiskDevices) > 1 && newLocalRootDevice["pool"] == "") {
+		return BadRequest(fmt.Errorf("Profiles specifiy conflicting pools for the containers root disk device."))
+	}
+
+	// The old local devices provided a root disk device and the new ones
+	// don't. So check whether the profiles provide a unique one.
+	if oldLocalRootDevice != nil && newLocalRootDevice == nil {
+		if len(profileRootDiskDevices) == 0 {
+			return BadRequest(fmt.Errorf("The change would leave the container without a root disk device."))
+		} else if profileRootDiskDevices[0] != oldLocalRootDevice["pool"] {
+			return BadRequest(fmt.Errorf("The update would change the containers' storage pool."))
+		}
+	} else if newLocalRootDevice["pool"] != "" {
+		if oldLocalRootDevice["pool"] != "" && oldLocalRootDevice["pool"] != newLocalRootDevice["pool"] {
+			return BadRequest(fmt.Errorf("The update would change the containers' storage pool."))
+		} else if oldLocalRootDevice["pool"] == "" && newLocalRootDevice["pool"] != profileRootDiskDevices[0] {
+			return BadRequest(fmt.Errorf("The update would change the containers' storage pool."))
+		}
+	}
```

Comment:
I don't get why we're adding all this complexity here instead of doing the checks in Update()

Doesn't doing all this here also means that Patch() isn't covered?